ENH: Clarifying merge and review practices

choldgraf opened the pull request:
This is a proposal to clarify and make more explicit our policies and guidelines around PR merging and reviewing. See issues below for discussion and (I think) suggested language. I tried to strike a balance between "very specific and rigid" and "so generic that it is hard to use this to make decisions".

- closes https://github.com/jupyterhub/team-compass/issues/465
- closes https://github.com/jupyterhub/team-compass/issues/594
- closes https://github.com/jupyterhub/team-compass/issues/466

choldgraf commented:
Thanks for the suggestions all - I've made a few edits to this one to address @consideRatio's comments here as well as a few in the issue referenced above!

manics commented on docs/practices/pull_requests.md:
```diff
@@ -0,0 +1,73 @@
+Our goals are to balance inclusive and transparent team practices with the reality that we are all asynchronous and don't want to pay a large coordination penalty for every change and decision.
+
+**All changes to a JupyterHub repository should happen via Pull Requests** and **each pull request should reference at least one issue**.
+Any exceptions to this should be explicitly listed in team documentation.
```
The only exception I can think of to the PR workflow is for making releases.

manics reviewed (commented):
Thanks for working on this.

The only controversial suggestion I see is "each pull request should reference at least one issue", I don't think this is something we've encouraged in the past.

choldgraf commented:
I felt that it was under the "I'm pretty sure we already implicitly follow this" category which is why I put it in. IMO (like everything else) it's not a strict rule (eg not needed for small dogs changes) but a good guideline

choldgraf commented:
I felt that it was under the "I'm pretty sure we already implicitly follow this" category which is why I put it in. IMO (like everything else) it's not a strict rule (eg not needed for small dogs changes) but a good guideline

minrk commented on docs/practices/pull_requests.md:
```diff
@@ -0,0 +1,73 @@
+Our team practices should encourage transparency and collaboration, and reviewing one another's work is a great way to ensure that we share context and knowledge, and improve the codebase.
+Our goals are to balance inclusive and transparent team practices with the reality that we are all asynchronous and don't want to pay a large coordination penalty for every change and decision.
+
+**All changes to a JupyterHub repository should happen via Pull Requests** and **each pull request should reference at least one issue**.
```
Yeah, I agree. I wouldn't want to imply that we want folks to open an Issue immediately before opening a PR that closes it. PRs _are_ issues, so I find having a discussion in one place for an issue and its solution is often preferable to having one discussion in two places. If I found a problem, I would only open an Issue first if I knew I didn't have time to open a PR instead, and I'd encourage others to do the same. I think that's easier to keep track of for ~everyone unless you are doing specific tracking/reporting based on closed Issues only, which we don't.

If folks want, we could mention the reasons for opening a PR against an already acknowledged issue, and the potential reasons to open an issue first to discuss a change before working on a solution. I mainly don't want anyone at the stage of 'I have my change, time to open a PR' deciding they need to open an _issue_ first for some reason, and I feel like this text is telling folks to they _should_ do that, when the opposite is true.

choldgraf replied:
Ah ok - I will remove this then. My goal in this PR is not to propose anything too new.

For what it's worth, I disagree. I think GitHub is already not very accessible to non-developers, and pull requests are the worst place to make decisions if you want to include non-developers.

I also think that doing all of the work in just a PR (for non-trivial changes) conflates two distinct parts of the process that require input from different stakeholders :

1. Agreeing on the problem to solve, agreeing that it is worth solving, and aligning on a general idea of what the solution looks like (I think both of these should be in issues or a discussion forum)
2. Agreeing on an _implementation_ to the problem and potentially clarifying design considerations that come up during implementation (i think this needs to be PRs)

Obviously I don't think that's a strict process for *all* changes, but for anything non-trivial, yes my preference is to always have an issue or forum topic for discussion first.

manics replied:
One concern I have is that we're struggling to stay on top of all issues and PRs across our repositories. Long-lasting issues and PRs are frustrating for contributors or users, as is the lack of well defined process in some cases, but that's _partially_ balanced by being an open-source project which engenders some goodwill and understanding of how open-source works.

That goodwill is lost if we ask contributors to follow a certain process but don't follow-up- if I'm interacting with a project with no guidelines it's frustrating but I understand that the voluntary nature of open-source means things can be a bit haphazard sometimes. It's a lot more frustrating to have guidelines which I've followed but which aren't reciprocated due to lack of time or resources.

choldgraf replied:
I think this is a good point. I will try and change the wording tomorrow to be more nuanced

betatim replied:
Big 👍  to not encouraging people to create issues a few seconds before they create a PR that closes it. For the reasons that Min and Simon articulated. I think a great strength of a PR is that you can have all the discussion and the change right there, basically all the context you need to take action (review, merge, etc) is right there.

On the topic of keeping discussions separate (should we even solve this? vs how to solve this?) I think doing it all in the PR discussion, especially if there is code to look at makes sense. In rare cases, for contentious changes it might make sense to move the discussion to an issue to try and separate it from a particular implementation (and its author). Basically a way to grease the social wheels of making progress with reduced emotional attachment to a particular change. But I think this is rare, so we can do it ad-hoc.

Sometimes the result of a PR might be "this is a problem, but somehow this doesn't feel like the right implementation and we don't know what the right implementation looks like/don't have time to work on it" - in these cases I think it makes sense to open an issue with a short summary of what happened in the PR to "reboot" the conversation and make it easier to get into it when revisiting it later.

choldgraf replied:
I've re-worked the language in the latest commit to be softer on this statement, can you focus discussion on that language to suggest edits etc? I'm not going to debate when we should/shouldn't open issues (my opinion is that we should always open issues, but I recognize I'm in the minority on that one and I don't feel the need to convince people here as long as the language in this PR is OK to y'all).

minrk replied:
Interesting! I don't view issues and PRs as distinguishable for conversation purposes in terms of participants or goals, other than PRs having concrete changes to discuss, while Issues don't. Forums/Discussions are definitely a mechanism for broadening participation. I've never had the impression that Issues have different visibility/participation than PRs.

Suggesting opening an Issue for discussion prior to PRs seems sensible. Maybe also suggesting soliciting feedback for large/user-facing changes via wider mediums like the forum even after opening a PR is a good idea, too.

choldgraf replied:
My thinking on this changed considerably once I started using GitHub as a collaboration platform with people who were not developers. In my experience, when a non-developer is asked to participate in a discussion, but then sees a page that displays a bunch of source code, they are often significantly discouraged to participate because they feel like they're out of their element.

I also find that putting the code in front of everybody's face tends to focus discussions on implementation-centric ideas, which are often very different from design- and UX-level ideas. Pull Requests are inherently a code-centric thing, and much of the activity around deciding "how should our tools change?" does not require understanding our codebase. In fact, many of the people out there with deep expertise in UX and product design are not coders at all! I think it's important for somebody to be able to be heard without requiring them to look at source code.

Ah I can feel myself slipping into a debate about this 😅 as I say I am happy to just stick with the minimal language that will let us move forward, and I'm happy to have deeper discussions about inclusion and accessibility in our team processes in follow-ups.

minrk replied:
Those are excellent points, and I don't think I appreciated the _presence_ of code being off-putting for folks otherwise happy with Issues. My main experience has been that Issues are _similarly exclusionary_ to PRs, not that PRs aren't exclusive. So in the goal of soliciting broader participation, I haven't found increasing the use of granular Issues improves things over discussion in PRs to be beneficial, whereas _other_ lower-traffic, non-code-centric places like the forum/projects/discussions/blogs/social media definitely is.

But I agree, we should always be working to improve our processes for inclusion, and that's a bigger, broader topic. Thanks for always being so thoughtful on this!

choldgraf commented:
OK I believe that I've addressed the comments above by removing the language about "must" open issues, and instead providing recommendations for how to be inclusive of others in discussion for more complex proposals.

betatim commented on docs/practices/pull_requests.md:
```diff
@@ -0,0 +1,79 @@
+  Go out of your way to make it easy for others to participate, and make your best effort at incorporating suggestions and feedback from others.
+  If a topic requires discussion, open an issue or a forum thread so that others can participate and align on an approach before diving into a code implementation.
+  We often save a lot of time by having design-level discussion before we get to the Pull Request.
+- **Link to discussions and context**. Almost any change should first be proposed in an issue, or in another space that is accessible to non-coders (like a forum post).
```
I'm not super happy with this. Without historical context this makes me think that we will end up with forum threads that should be issues, the problem being that forum threads can't easily be linked from other issues/PRs.

A proposal that I am more happy with, but also not super happy:

Let's add something to the previous item (include others) about shopping your idea around and maybe how to judge when you should do that/where you should do that. Maybe the rule of thumb is "the more people are impacted by your change, the broader you should shop around your proposal"? - but it might be hard to know how many people are impacted by your change. An alternative is "the more fundamental your change..." - but that kinda has the same problem (how do you know?).

With the above addition I would delete the first sentence here ("Almost any change ... forum post).". That way this item stays on the topic of "you should link to discussions and context". Creating an issue/forum post is a way of creating that context, which is different from then linking to it.

GeorgianaElena commented on docs/practices/pull_requests.md:
```diff
@@ -0,0 +1,79 @@
+  Go out of your way to make it easy for others to participate, and make your best effort at incorporating suggestions and feedback from others.
+  If a topic requires discussion, open an issue or a forum thread so that others can participate and align on an approach before diving into a code implementation.
+  We often save a lot of time by having design-level discussion before we get to the Pull Request.
+- **Link to discussions and context**. Almost any change should first be proposed in an issue, or in another space that is accessible to non-coders (like a forum post).
+  A Pull Request should link to (and optionally close) any other locations where discussion happened.
+  This gives each change more provenance and separates design from implementation.
```
Hmm, maybe we can melt this one with the first bullet (see my suggestion about renaming it to `provide context`) to reduce any controversy?

Maybe something like:
```suggestion
- **Provide context**. Provide enough discussion and context sharing so that every team member understands the rationale for a change, even if they may not personally agree with it.
  A Pull Request should link to (and optionally close) any locations where discussion happened.
  This can give each change more provenance and separates design from implementation.
  The more complex or impactful a change is, the more time should be given to discuss and include the perspective of others. These discussions can happen in issues, pull requests or in another space that is accessible to non-coders (like a forum post). There aren't any strict rules about how to choose the discussion space, as this usually depends on the type, scope and size of change that's being proposed. So use your best judgement when choosing one.
```

GeorgianaElena commented on docs/practices/pull_requests.md:
```diff
@@ -0,0 +1,79 @@
+  Use your best judgment for how long to wait to ensure inclusive decision making process and be sure to [include others in discussion](prs:principles).
+  The more complex, the more time you should give people.
+  The more people agree, the more quickly you can safely merge.
+- **Changes to team policies** may be merged with at least two approvals from {team}`JupyterHub Council` members, no objections from any team member, and has been open for around five working days (follow the same guidelines as above).
```
```suggestion
- **Changes to team policies** may be merged with **at least two approvals** from {team}`JupyterHub Council` members, no objections from any team member, and has been open for around five working days (follow the same guidelines as above).
```

GeorgianaElena commented on docs/practices/pull_requests.md:
```diff
@@ -0,0 +1,79 @@
+  If you approve the PR, go ahead and merge it!
+  If you don't immediately merge it, provide a rationale for why not.
+  If somebody else approves your PR and you think it's safe to merge, feel free to self-merge.
+- **Complex or potentially controversial PRs** may be merged with at least one approval from a {team}`Maintainer Team` member, no objections from any team member, and after has been open for around five working days.
```
```suggestion
- **Complex or potentially controversial PRs** may be merged with at least one approval from a {team}`Maintainer Team` member, no objections from any team member, and after has been open for at least five working days.
```

GeorgianaElena commented on docs/practices/pull_requests.md:
```diff
@@ -0,0 +1,79 @@
+  us for it later.
+- **There is documentation**. If we're changing or adding functionality, ensure that there is documentation describing the change to others.
+  Functionality that isn't documented is much less discoverable and maintainable.
+- **The code is secure**. We must avoid introducing changes that may lead to a security vulnerability. If your code is complicated please include comments to ensure future maintainers do not inadvertently introduce a new vulnerability.
```
I'm not sure, but from this diff view it looks there's a bit of inconsistency in this section around splitting by sentence and splitting by line length?

```suggestion
- **The code is secure**. We must avoid introducing changes that may lead to a security vulnerability.
  If your code is complicated please include comments to ensure future maintainers do not inadvertently introduce a new vulnerability.
```

GeorgianaElena commented on docs/practices/pull_requests.md:
```diff
@@ -0,0 +1,79 @@
+(prs:principles)=
+## Principles to follow
+
+- **Don't surprise people**. We pay a team penalty when somebody changes something that wouldn't be expected, especially if it is unwelcome.
+  Provide enough discussion and context sharing so that every team member understands the rationale for a change, even if they may not personally agree with it.
+  The more complex or impactful a change is, the more time should be given to discuss and include the perspective of others.
```
See also the comment below about merging this bullet point with the one about linking issues https://github.com/jupyterhub/team-compass/pull/596#discussion_r1039449483

manics commented:
I think the important thing about https://simonwillison.net/2022/Oct/29/the-perfect-commit/ is that there's a description for what the issue should cover (Background, Links to things, Code snippets illustrating potential designs and false-starts, Decisions, Screenshots, Prototypes), and it refers to linking commits to issues.

As far as I can tell linking a single commit to an issue is the same as a pull request with commits, the only difference being that in the former the commit is linked to an issue through the commit message whereas in the latter it's automatically part of the PR.

choldgraf commented:
I took another pass at the text here after feedback from you all, and also incorporated the suggestions from @betatim and @GeorgianaElena. I tried to consolidate a little bit per the suggestions above. What do folks think about it now?

betatim commented on docs/practices/pull_requests.md:
```diff
@@ -0,0 +1,84 @@
+  Actively invite participation and comments from others, especially if you think they are interested in the proposed change.
+  Go out of your way to make it easy for others to participate, and make your best effort at incorporating suggestions and feedback from others.
+  For discussions that are complex, potentially controversial, or have an uncertain resolution, open an issue or forum post to discuss with others before diving into a code implementation.
+  Consider the skills and comfort level of others in the discussion, and choose a location and format for the discussion that will be accessible to all relevant stakeholders.
```
Not sure if this is equivalent :-/ It is an attempt at saying "We want you to want others to take part in discussions, think about where the people with the skills are, go there. Don't expect them to come to you."
```suggestion
  Consider the skills and comfort level of the people who you want to take part in the discussion. Choose a location and format for the discussion that will be accessible to them. Think about where those with expertise on the topic can be reached, not where is easiest for you.
```

betatim commented on docs/practices/pull_requests.md:
```diff
@@ -0,0 +1,84 @@
+
+- **It's quality code**.
+  We know this is somewhat subjective, but ensure that the code is well-organized and thoughtfully-written, that any new features are documented, and that it abides by best-practices in Python, JavaScript, etc.
+  Do not nit-pick stylistic preferences, but if you genuinely think there's a clear way to improve the code, please make suggestions.
```
I like this and wonder if we should expand it or add another line to say "don't nitpick" or "mark your nits as nits". I think a positive way to phrase it is "Try and stay big picture, worry about details later." Kinda "is this a change we want to make?", "is this a good way of doing it?", and then at the very end "should this loop variable be `n` or `i`?". I think I (and many others) are easily sucked into focusing on the details line-by-line of a PR instead of taking the step back. And often while you are looking at something you realise that "adjacent code" is not nice as well and could do with some tweaks and then ask for them. Instead I think we should either make an issue, or clearly mark such comments as "something for a new PR". All to try and not broaden the scope of a PR.

choldgraf replied:
I like the way that [consent-based decision-making often describes it](https://www.sociocracyforall.org/consent-decision-making/).

Basically, there is "what you want", then there is "what you will tolerate", and then there is "what you will not tolerate". We want people to approve PRs within "what they will tolerate", and we want people only to block merging if something has a major negative strategic implication on the project (after one or two rounds of expected minor improvements).

betatim replied:
See my comment about "merge because this is not worse than what we have". So I agree with your "merge what you can tolerate".

What I was trying to get at with my comment was that you should try and start your review with "big picture" thinking. And then later go line by line. And even later go character by character. Or maybe decide somewhere along the line, before reaching the end of this "zooming in", that this is super good enough and merge it.

betatim replied:
Marking this as resolved because of https://github.com/jupyterhub/team-compass/commit/3064410a420a6ba8e38ec372c5988a622390c206

choldgraf commented:
@betatim I added one more commit to try and capture what you said, I put it at the top of the list because I actually think it is the most important. WDYT?

https://github.com/jupyterhub/team-compass/pull/596/commits/3064410a420a6ba8e38ec372c5988a622390c206

manics reviewed (approved):
Looks a lot nicer now!

Not sure where best to mention this in the doc, but it might be worth mentioning breaking changes as a class of PRs requiring more consideration since they can't be easily reversed, even if it's a small code change?

consideRatio reviewed (approved):
@choldgraf wieee amazing work to tie together so many peoples thoughts!!!! :heart: :tada: :sunflower:!!

betatim commented:
RTD build failed with " fatal: reference is not a tree: 3064410a420a6ba8e38ec372c5988a622390c206 ". Trying to figure out if I can restart it somehow. However we saw a RTD build fail in a different JH repo for "no reason", so maybe this is due to something on the RTD being broken at the moment.

betatim commented:
Merged. The comment about backwards incompatible changes is a good one, in the interest of keeping people moving along, let's put it in a new PR.

choldgraf commented:
Yay! Thanks all for the helpful feedback :-)
